### ISSUE TITLE
VIDIOC_G_FMT possibly causes VIDIOC_REQBUFS to return unusable buffers and thus mmap( ) fails

### ISSUE THREAD

**Issue by periwinklepreacher** [mpromonet/v4l2rtspclient#2]
V4l2 kernel documentation shows that VIDIOC_REQBUFS asks for a specific number of buffers then verifies that buffers are actually allocated by kernel and that the number is sufficient for correct operation. 

V4l2MmapDevice::start( ) sets requested buffers to zero. Confusing since V4l2MmapDevice::stop( ) uses similar logic to release allocated buffers. The following, based on reference code, fails to allocate any buffers when requested buffer count is zero:

> reqbuf: reqbuf.c:31: main: Assertion `reqbuf.count > 0' failed.

Assigning reqbuf.count a value of 10 results in a subsequent failure:

> reqbuf: reqbuf.c:46: main: Assertion `buffer.length >= buffer.bytesused' failed.

Maybe some sort of environment failure in Linux Mint 19.1 Tessa 4.15.0-45-generic is the problem? Unclear. Not certain how v4l2rtspclient is meant to use a buffer when buffer.length < buffer.bytesused'. An orderly shutdown of v4l2rtspclient with an error message could be helpful when diagnosing this situation. 

```
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <assert.h>
#include <errno.h>
#include <sys/mman.h>
#include <sys/ioctl.h>
#include <linux/videodev2.h>

void main( ) {
	struct v4l2_requestbuffers reqbuf;
	int i, fd;

	fd = open( "/dev/video0", O_RDWR | O_NONBLOCK );
	assert( fd > 0 );

	memset( &reqbuf, 0, sizeof( reqbuf ) );
	reqbuf.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
	reqbuf.memory = V4L2_MEMORY_MMAP;
//	reqbuf.count = 10;

	if( -1 == ioctl ( fd, VIDIOC_REQBUFS, &reqbuf ) ) {
		if ( errno == EINVAL )
			printf( "Video capturing or mmap-streaming is not supported\n" );
		else
			perror( "VIDIOC_REQBUFS" );
		exit( EXIT_FAILURE );
	}
	assert( reqbuf.count > 0 );

	for( i = 0; i < reqbuf.count; i++ ) {
		struct v4l2_buffer buffer;

		memset( &buffer, 0, sizeof( buffer ) );
		buffer.type = reqbuf.type;
		buffer.memory = V4L2_MEMORY_MMAP;
		buffer.index = i;

		if( -1 == ioctl( fd, VIDIOC_QUERYBUF, &buffer ) ) {
			perror( "VIDIOC_QUERYBUF" );
			exit( EXIT_FAILURE );
		}

		assert( buffer.length >= buffer.bytesused );
	}

	close( fd );
}
```

**Comment by mpromonet** [comment]
Hi Nikole,

I am not sure to follow you, in [V4l2MmapDevice::start](https://github.com/mpromonet/libv4l2cpp/blob/ea5e163b2761ecff9897b604ba8dcafbb365de72/src/V4l2MmapDevice.cpp)  

	struct v4l2_requestbuffers req;
	memset (&req, 0, sizeof(req));
	req.count               = V4L2MMAP_NBBUFFER;
	req.type                = m_deviceType;
	req.memory              = V4L2_MEMORY_MMAP;

	if (-1 == ioctl(m_fd, VIDIOC_REQBUFS, &req)) 

As V4L2MMAP_NBBUFFER is defined to be 10 in https://github.com/mpromonet/libv4l2cpp/blob/ea5e163b2761ecff9897b604ba8dcafbb365de72/inc/V4l2MmapDevice.h, the request is done with 10.

Surely this code could be improve,  but I am not catching what you propose.
You can submit a pull request to clarify this.
Best Regards,
Michel.

**Comment by periwinklepreacher** [comment]
The issue with req.count is simply because my brain refuses to recognize your line of code. ;-)
> req.count = V4L2MMAP_NBBUFFER

The buffer count is assigned exactly as required by the API. My apologies for saying otherwise.

As for the second issue I have identified the workflow where 0 length buffers are returned to v4l2rtspclient by VIDIOC_REQBUFS. Discrepancy in call to VIDIOC_G_FMT, pixelformat=BGR4 vs pixelformat=H264 , appears to highlight the problem. See logs below.

TL;DR
VIDIOC_G_FMT possibly causes VIDIOC_REQBUFS to return unusable buffers and thus mmap( ) fails.

**Workaround**
```
#!/bin/bash
sudo modprobe -r v4l2loopback
sudo modprobe v4l2loopback
v4l2rtspclient "$@"
```
Confirmed that vlc can render H264 stream from v4l2rtspclient 4vl2loopback device under specific  conditions. v4l2rtspclient is a nice light implementation compared to using ffmpeg for the same purpose. Excellent and thank you.

**v4l2loopback kernel log after modprobe reload**
> kernel: [39428.182983] v4l2-loopback[1844]: opened dev:00000000731c3667 with image:          (null)
> kernel: [39428.182986] v4l2loopback.c:1845[v4l2_loopback_open]
> kernel: [39428.182998] video0: VIDIOC_QUERYCAP: driver=v4l2 loopback, card=Dummy video device (0x0000), bus=platform:v4l2loopback-000, version=0x00040f12, capabilities=0x85208003, device_caps=0x85208003
> kernel: [39428.184942] v4l2loopback.c:907[vidioc_g_fmt_out]
> kernel: [39428.184945] video0: VIDIOC_G_FMT: type=vid-out, width=0, height=0, pixelformat=BGR4, field=none, bytesperline=0, sizeimage=0, colorspace=8, flags=0x0, ycbcr_enc=0, quantization=0, xfer_func=0
> kernel: [39428.186044] v4l2loopback.c:984[vidioc_s_fmt_out]
> kernel: [39428.186048] v4l2loopback.c:930[vidioc_try_fmt_out]
> kernel: [39428.186051] v4l2-loopback[949]: trying image 640x480
> kernel: [39428.186054] v4l2-loopback[989]: s_fmt_out(0) 0...1228800
> kernel: [39428.186056] v4l2-loopback[992]: outFOURCC=H264
> kernel: [39428.186060] v4l2loopback.c:1977[allocate_buffers]
> kernel: [39428.186063] v4l2-loopback[1998]: allocating 9830400 = 1228800x8
> kernel: [39428.187081] v4l2-loopback[2006]: vmallocated 9830400 bytes
> kernel: [39428.187084] v4l2loopback.c:2007[allocate_buffers]
> kernel: [39428.187087] v4l2loopback.c:2019[init_buffers]
> kernel: [39428.187092] v4l2loopback.c:2045[init_buffers]
> kernel: [39428.187095] video0: VIDIOC_S_FMT: type=vid-out, width=640, height=480, pixelformat=H264, field=none, bytesperline=0, sizeimage=1228800, colorspace=8, flags=0x0, ycbcr_enc=0, quantization=0, xfer_func=0
> kernel: [39428.187210] v4l2loopback.c:1363[vidioc_reqbufs]
> kernel: [39428.187213] v4l2-loopback[1368]: reqbufs: 1	10=8
> kernel: [39428.187216] v4l2loopback.c:2019[init_buffers]
> kernel: [39428.187220] v4l2loopback.c:2045[init_buffers]
> kernel: [39428.187221] video0: VIDIOC_REQBUFS: count=8, type=vid-out, memory=mmap
> kernel: [39428.187278] v4l2loopback.c:1434[vidioc_querybuf]
> kernel: [39428.187281] v4l2-loopback[1455]: buffer type: 1 (of 8 with size=1228800)
> kernel: [39428.187283] video0: VIDIOC_QUERYBUF: 10:57:03.00722285 index=0, type=vid-out, flags=0x00000002, field=none, sequence=0, memory=mmap, bytesused=1228800, offset/userptr=0x0, length=1228800

**v4l2rtspclient log after modprobe reload**
> [NOTICE] src/V4l2Device.cpp:208
> 	/dev/video0:H264 size:640x480 bufferSize:1228800
> [NOTICE] src/V4l2MmapDevice.cpp:71
> 	Device /dev/video0 nb buffer:8
> [INFO] src/V4l2MmapDevice.cpp:90
> 	Device /dev/video0 buffer idx:0 size:1228800

**v4l2loopback kernel log after v4l2rtspclient restart (without v4l2loopback module reload)**
kernel: [39708.212946] video0: VIDIOC_G_FMT: type=vid-out, width=640, height=480, pixelformat=H264, field=none, bytesperline=0, sizeimage=1228800, colorspace=8, flags=0x0, ycbcr_enc=0, quantization=0, xfer_func=0
kernel: [39708.212982] v4l2loopback.c:1363[vidioc_reqbufs]
kernel: [39708.212985] v4l2-loopback[1368]: reqbufs: 1	10=8
kernel: [39708.212988] v4l2loopback.c:2019[init_buffers]
kernel: [39708.212992] v4l2loopback.c:2045[init_buffers]
kernel: [39708.212994] video0: VIDIOC_REQBUFS: count=8, type=vid-out, memory=mmap
kernel: [39708.213017] v4l2loopback.c:1434[vidioc_querybuf]
kernel: [39708.213059] v4l2-loopback[1455]: buffer type: 1 (of 8 with size=0)
kernel: [39708.213119] video0: VIDIOC_QUERYBUF: 11:01:43.00637122 index=0, type=vid-out, flags=0x00000002, field=none, sequence=0, memory=mmap, bytesused=1228800, offset/userptr=0x0, length=0

**v4l2rtspclient log after v4l2rtspclient restart (without v4l2loopback module reload)**
[DEBUG] src/v4l2writer.cpp:61
	NAL:7
[NOTICE] src/v4l2writer.cpp:67
	geometry:640x480
[NOTICE] src/V4l2Device.cpp:117
	driver:v4l2 loopback capabilities:85208003 mandatory:4000002
[NOTICE] src/V4l2Device.cpp:119
	/dev/video0 support output
[NOTICE] src/V4l2Device.cpp:120
	/dev/video0 support capture
[NOTICE] src/V4l2Device.cpp:122
	/dev/video0 support read/write
[NOTICE] src/V4l2Device.cpp:123
	/dev/video0 support streaming
[NOTICE] src/V4l2Device.cpp:147
	/dev/video0:H264 size:640x480
[NOTICE] src/V4l2MmapDevice.cpp:71
	Device /dev/video0 nb buffer:8
[INFO] src/V4l2MmapDevice.cpp:90
mmap: Invalid argument
	Device /dev/video0 buffer idx:0 size:0

**Comment by periwinklepreacher** [comment]
Successfully tested fix using shell script workaround. 

**Workaround**

```
#!/bin/bash
sudo modprobe -r v4l2loopback
sudo modprobe v4l2loopback
v4l2rtspclient "$@"
```
